File: Source/src_ios/StrengthTester.swift
//
// Copyright 2018 Wultra s.r.o.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions
// and limitations under the License.
//

import Foundation
import PasswordMeter

/// Class that provides functionality for testing strngth of passwords and PINs.
///
/// Warning: Only one insttance of this class can exist. Init will return nil if you'll try to create 2nd instance.
public class StrengthTester {
    
    private static weak var instance: StrengthTester?
    
    /// Will create instance of the tester. Only one instance can be created, otherwise will return nil.
    /// If dictionary setup fails, init will also return nil.
    ///
    /// - Parameter dictionary: language dictionary for password checking
    public init?(dictionary: StrengthTesterDictionary? = nil) {
        
        guard StrengthTester.instance == nil else {
            print("[StrengthTester] WARNING: Only one instance of StrengthChecker can exist")
            return nil
        }
        
        if let dictionary = dictionary {
            guard
                let result = dictionary.path.data(using: .utf8)?.withUnsafeBytes({ WPM_setPasswordDictionary($0) }),
                result == 1 else {
                    print("[StrengthTester] WARNING: Instance with give dict cannot be configured.")
                    return nil
            }
        }
        
        StrengthTester.instance = self
    }
    
    deinit {
        WPM_freePasswordDictionary()
        StrengthTester.instance = nil
    }
    
    
    /// Tests strength of the PIN.
    ///
    /// - Parameter pin: PIN to evaluate. This needs to be digits only.
    /// - Returns: Optionset of found issues.
    public func testPin(_ pin: String) -> PinTestResult {
        
        guard Int(pin) != nil else {
            return .pinFormatError
        }
        
        return pin.data(using: .utf8)!.withUnsafeBytes({ (ptr: UnsafePointer<Int8>) -> PinTestResult in
            
            let result = WPM_testPasscode(ptr)
            
            var cr: PinTestResult = []
            
            if result.rawValue & NOT_UNIQUE_WPM.rawValue != 0 {
                cr.insert(.notUnique)
            }
            
            if result.rawValue & HAS_PATTERN_WPM.rawValue != 0 {
                cr.insert(.patternFound)
            }
            
            if result.rawValue & REPEATING_CHARACTERS_WPM.rawValue != 0 {
                cr.insert(.repeatingCharacters)
            }
            
            if result.rawValue & POSSIBLY_DATE_WPM.rawValue != 0 {
                cr.insert(.possiblyDate)
            }
            
            if result.rawValue & FREQUENTLY_USED_WPM.rawValue != 0 {
                cr.insert(.frequentlyUsed)
            }
            
            if cr.isEmpty {
                cr.insert(.ok)
            }
            
            return cr
        })
    }
    
    /// Tests strength of the password.
    ///
    /// Note that result is affected by dictionary that was set in `init` method.
    ///
    /// - Parameter password: Password to test
    /// - Returns: Strength of the password
    public func testPassword(_ password: String) -> PasswordStrength {
        guard password.count > 0 else { return .veryWeak }
        return PasswordStrength(password.data(using: .utf8)!.withUnsafeBytes{ return WPM_testPassword($0) })
    }
}

/// Specific language dictionary for password strength checking. Note that implementation for each language is provided in separated podspec.
public struct StrengthTesterDictionary {
    let path: String
    init(_ path: String) {
        self.path = path
    }
}

/// Result of PIN testing
public struct PinTestResult: OptionSet {
    
    public let rawValue: Int
    
    public init(rawValue: Int) {
        self.rawValue = rawValue
    }
    
    /// No issues found
    public static let ok = PinTestResult(rawValue: 1 << 0)
    /// Not enough unique digits found
    public static let notUnique = PinTestResult(rawValue: 1 << 1)
    /// Too much repeating characters
    public static let repeatingCharacters = PinTestResult(rawValue: 1 << 2)
    /// There is a pattern in this pin (for example 1357)
    public static let patternFound = PinTestResult(rawValue: 1 << 3)
    /// Tested pin could be date (for example 2512 as birthday - 25th of december)
    public static let possiblyDate = PinTestResult(rawValue: 1 << 4)
    /// Tested pin is in TOP used pins (like 1234 as number 1 used pin)
    public static let frequentlyUsed = PinTestResult(rawValue: 1 << 5)
    /// PIN contains non-digit characters
    public static let pinFormatError = PinTestResult(rawValue: 1 << 6)
}

/// Result of password strength test
public enum PasswordStrength {
    case veryWeak
    case weak
    case modetrate
    case good
    case strong
    
    fileprivate init(_ score: WPM_password_check_score) {
        switch score {
        case VERY_WEAK_PASSWORD_SCORE_WPM: self = .veryWeak
        case WEAK_PASSWORD_SCORE_WPM: self = .weak
        case MODERATE_PASSWORD_SCORE_WPM: self = .modetrate
        case GOOD_PASSWORD_SCORE_WPM: self = .good
        case STRONG_PASSWORD_SCORE_WPM: self = .strong
        default: fatalError("Uncovered password check case", file: "StrengthTester.swift")
        }
    }
}

File: Source/src_android/PinCheckerLibrary/src/main/java/com/wultra/utils/StrengthTester.java
package com.wultra.utils;

import android.content.res.AssetManager;
import android.support.annotation.Nullable;

import java.util.EnumSet;

/**
 *  Class that provides methods for testing strength of passwords and PINs. Only once instance can be created at once!
 */
public class StrengthTester {

    static {
        System.loadLibrary("StrengthTester");
    }

    private native int init(AssetManager manager, String dictionaryAsset);
    private native int deinit();

    private native int testPasswordJNI(String password);
    private native int testPinJNI(String password);

    private static StrengthTester instance;

    /**
     * Creates instance. Note that only one instance can created, otherwise exception will be thrown.
     * @param manager Asset manager where asset is present
     * @param dictionaryAsset name of the language dictionary in asset manager
     */
    public StrengthTester(@Nullable AssetManager manager, @Nullable String dictionaryAsset) throws SingleInstanceExteption {

        if (instance != null) {
            throw new SingleInstanceExteption();
        }

        init(manager, dictionaryAsset);
        instance = this;
    }

    @Override
    protected void finalize() throws Throwable {
        super.finalize();
        freeDictionary();
    }

    /**
     * Releases dictionary. Call this when instance is no longer needed.
     */
    public void freeDictionary() {
        deinit();
    }

    /**
     * Tests strength of the password. Result can ne affected by dictionary given in construtor.
     * @param password Password to test
     * @return Strength of the password
     */
    public PasswordStrength testPassword(String password) {
        switch (testPasswordJNI(password)) {
            case 0: return PasswordStrength.VERY_WEAK;
            case 1: return PasswordStrength.WEAK;
            case 2: return PasswordStrength.MODERATE;
            case 3: return PasswordStrength.GOOD;
            default: return PasswordStrength.STRONG;
        }
    }

    /**
     * Scans PIN for possible issues.
     * @param pin PIN to scan.
     * @return Set of issues.
     */
    public EnumSet<PinResultFlag> testPin(String pin) {

        int result = testPinJNI(pin);

        EnumSet<PinResultFlag> set = EnumSet.noneOf(PinResultFlag.class);

        if ((result & 1) != 0) {
            set.add(PinResultFlag.OK);
        }

        if ((result & 2) != 0) {
            set.add(PinResultFlag.NOT_UNIQUE);
        }

        if ((result & 4) != 0) {
            set.add(PinResultFlag.REPEATING_CHARACTERS);
        }

        if ((result & 8) != 0) {
            set.add(PinResultFlag.HAS_PATTERN);
        }

        if ((result & 16) != 0) {
            set.add(PinResultFlag.POSSIBLY_DATE);
        }

        if ((result & 32) != 0) {
            set.add(PinResultFlag.FREQUENTLY_USED);
        }

        return set;
    }

    /**
     * Result of PIN testing
     */
    public enum PinResultFlag {
        /**
         * Passcode is OK, no issues found
         */
        OK,
        /**
         * Passcode doesn't have enough unique digits
         */
        NOT_UNIQUE,
        /**
         * There is significant amount of repeating characters in the passcode
         */
        REPEATING_CHARACTERS,
        /**
         * Repeating pattern was found in the passcode
         */
        HAS_PATTERN,
        /**
         * This passcode can be date (and possible birthday of the user)
         */
        POSSIBLY_DATE,
        /**
         * Passcode is in most used passcodes
         */
        FREQUENTLY_USED
    }

    /**
     * Result of password strength test
     */
    public enum PasswordStrength {
        VERY_WEAK,
        WEAK,
        MODERATE,
        GOOD,
        STRONG
    }

    public static class SingleInstanceExteption extends Exception {

    }
}